code/modules/mob/living/living_blood.dm

```diff
@@ -0,0 +1,34 @@
+//Transfers blood from container to vessels
+/mob/living/proc/inject_blood(var/amount, var/datum/reagents/donor)
+	var/decl/species/my_species = get_species()
+	if(!my_species?.blood_volume)
+		return //Don't divide by 0
+	var/injected_data = REAGENT_DATA(donor, my_species.blood_reagent)
+	var/chems = LAZYACCESS(injected_data, "trace_chem")
+	for(var/C in chems)
+		src.reagents.add_reagent(C, (text2num(chems[C]) / my_species.blood_volume) * amount)//adds trace chemicals to owner's blood
+
+/mob/living/get_blood_data()
+
+	var/data = ..()
+	data["blood_DNA"]  = get_unique_enzymes()
+	data["blood_type"] = get_blood_type()
+	data["species"]    = get_species_name()
```

Comment:
worried that for mobs without a species, this would result in `list("species" = null)` and similar. maybe `get_species_name() || "Unknown"`? or a new proc like `get_species_display_name()` which returns `"Unknown"` if species is null?